tests/unit/no-top-level-side-effects.test.ts

```diff
@@ -653,6 +653,72 @@ const valid: RuleTester.ValidTestCase[] = [
       code: `const u04 = ~a;`,
       options: [options.allowDerived]
     }
+  ],
+  ...[
+    {
+      code: `const x = { foo: ok() };`,
+      options: [{ allowedCalls: ['ok'] }],
+    },
+    {
+      code: `const x = {};`,
+      options: [],
+    },
+    {
+      code: `const x = { foo: 123 };`,
+      options: [],
+    },
+    {
+      code: `const x = { foo: 123, bar: 'baz' };`,
+      options: [],
+    },
+    {
+      code: `const x = { foo: () => 123 };`,
+      options: [],
+    },
+    {
+      code: `const x = { foo: require('./config') };`,
+      options: [options.commonjs],
+    },
+    {
+      code: `const x = { foo: (function(){ return 1 })() };`,
+      options: [options.allowIIFE]
+    },
```

Comment:
I would say this is happening at the top level - the function is called as soon as the module is evaluated/imported - so it's a side effect, position of the IIFE shouldn't matter. if we allow it via `allowIIFE` I would expect any IIFE that's called as a side effect would be allowed.

Reply:
actually I think maybe this is only valid when both alloweIIFE and allowDerived are set? thoughts?

Reply:
I suggest we revert the change in `no-top-level-side-effects.ts` for now and resolve this separately (you can adapt the test case or remove it altogether if you want).

Like I said, the option is a bit unclear and honestly here just because the [original project always allowed IIFEs](https://github.com/eramdam/eslint-plugin-top/blob/de3072d8e1aa41b636eb1c211c1d8460483007b6/lib/rules/noTopLevelSideEffect.ts#L48). My intend for it is to allow IIFEs at the actual top-level only and not anywhere else.

Reply:
that's fair, let me revert this.

Reply:
in that case though - are you saying this would always result in a violation with no options to override it

```
const x = { foo: (function(){ return 1 })() };
```

Reply:
> in that case though - are you saying this would always result in a violation with no options to override it
> 
> ```
> const x = { foo: (function(){ return 1 })() };
> ```
> 

Yes

Reply:
I created https://github.com/ericcornelissen/eslint-plugin-top/issues/1108 to resolve this.